sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystemProvider.java

```diff
@@ -221,10 +228,111 @@ public InputStream newInputStream(Path path, OpenOption... options) throws IOExc
+     *
+     * {@inheritDoc}
+     */
+    @Override
+    public OutputStream newOutputStream(Path path, OpenOption... options) throws IOException {
+        // If options are empty, add Create, Write, TruncateExisting as defaults per nio docs.
+        if (options == null || options.length == 0) {
+            options = new OpenOption[] {
+                StandardOpenOption.CREATE,
+                StandardOpenOption.WRITE,
+                StandardOpenOption.TRUNCATE_EXISTING };
+        }
+        List<OpenOption> optionsList = Arrays.asList(options);
+
+        // Check for unsupported options.
+        List<OpenOption> supportedOptions = Arrays.asList(
+            StandardOpenOption.CREATE_NEW,
+            StandardOpenOption.CREATE,
+            StandardOpenOption.WRITE,
+            // Though we don't actually truncate, the same result is achieved by overwriting the destination.
+            StandardOpenOption.TRUNCATE_EXISTING);
+        for (OpenOption option : optionsList) {
+            if (!supportedOptions.contains(option)) {
+                throw new UnsupportedOperationException("Unsupported option: " + option.toString());
+            }
+        }
+
+        // Write and truncate must be specified
+        if (!optionsList.contains(StandardOpenOption.WRITE)
+            || !optionsList.contains(StandardOpenOption.TRUNCATE_EXISTING)) {
+            throw new IllegalArgumentException("Write and TruncateExisting must be specified to open an OutputStream");
+        }
+
+        AzureResource resource = new AzureResource(path);
+        DirectoryStatus status = resource.checkDirStatus();
+
+        // Cannot write to a directory.
+        if (DirectoryStatus.isDirectory(status)) {
+            throw LoggingUtility.logError(logger, new IOException("Cannot open an OutputStream to a directory. Path: "
+                + path.toString()));
+        }
+
+        // Writing to an empty location requires a create option.
+        if (status.equals(DirectoryStatus.DOES_NOT_EXIST)
+            && !(optionsList.contains(StandardOpenOption.CREATE)
+            || optionsList.contains(StandardOpenOption.CREATE_NEW))) {
+            throw LoggingUtility.logError(logger, new IOException("Writing to an empty location requires a create "
+                + "option. Path: " + path.toString()));
+        }
+
+        // Cannot write to an existing file if create new was specified.
+        if (status.equals(DirectoryStatus.NOT_A_DIRECTORY) && optionsList.contains(StandardOpenOption.CREATE_NEW)) {
+            throw LoggingUtility.logError(logger, new IOException("A file already exists at this location and "
+                + "CREATE_NEW was specified. Path: " + path.toString()));
+        }
+
+        // Create options based on file system config
+        AzureFileSystem fs = (AzureFileSystem) (path.getFileSystem());
+        Integer blockSize = fs.getBlockSize() == null ? null : fs.getBlockSize().intValue();
```

Comment:
blockSize is long now right?

Reply:
Or is this on purpose due to constraints below?

Reply:
It's long in 73, which is why I made the config value a long in anticipation of that, but this branch doesn't have those updates yet, so I'll strike these conversions to int once all that gets merged in.